Mod request decode trace

TrainCoffeeBoy opened the pull request:
### Requestの修正
以下の３点の修正をしました。unit_test, integrate_testを実施しています。ft_testerもOKです。

**1.  chunkで0が来たらすぐにparseを終える（ft_testerの仕様に合わせた）
2.  URLデコードをURI, queryに適用
3.  TRACEを実装**
 - headerのparseが終わった時点で、body_にbuf_の内容を突っ込むようにしました。
 - content-type: message/http をresponseするようにRFCに明記あり、関係する関数をいじりました
 - TRACEで禁止されるheaderの組み合わせはRFC上には明記なく、content-length, chunkedなどをそのままheaderとしては取得し、bodyは無視するようにしました(APACHEではエラーを出す実装になっていましたが）
 
確認、よろしくお願いします。

実行例
`./nginDX configfiles/trace.conf `

```
:$ curl -i -X TRACE localhost:8888
HTTP/1.1 200 OK
Server: nginDX
Date: Fri, 30 Apr 2021 14:41:02 GMT
Content-Length: 80
Connection: keep-alive
Content-Type: message/http

TRACE / HTTP/1.1
Host: localhost:8888
User-Agent: curl/7.65.3
Accept: */*

```

Commit:
mod chunked action of 0

Commit:
ft_tester ok with POST test

Commit:
url decode ok

Commit:
trace ok with curl but not yet with test

Commit:
trace integration and unit test ok

Commit:
unit test with content-length and chunked ok

Commit:
merge remote origin

Commit:
trace.conf

Commit:
merge remote origin

dai65527 commented on test/integration_test/my_tester/main.rb:
```diff
@@ -250,4 +250,31 @@
+testcase.expectedResponseHeaderExistance.push "Date"
+testcase.add
+
+#####
```
パーセントエンコーディングのテスターも追加お願いいたします！

```ruby
casename = "get /ファイル名.html in percent encoding"
testcase = WebservTestCase.new casename, "127.0.0.1", 8000, "/%E3%83%95%E3%82%A1%E3%82%A4%E3%83%AB%E5%90%8D.html", "GET"
testcase.expectedCode = "200"
testcase.expectedBody = "<h1>日本語でおk</h1>\n".force_encoding("ASCII-8BIT")  # need to change encoding to compare results
testcase.expectedResponseHeader["Server"] = "nginDX"
testcase.expectedResponseHeader["Content-Length"] = testcase.expectedBody.length.to_s
testcase.expectedResponseHeader["Connection"] = "keep-alive"
testcase.expectedResponseHeader["Content-Type"] = "text/html"
testcase.expectedResponseHeaderExistance.push "Date"
testcase.expectedResponseHeaderExistance.push "Last-Modified"
testcase.add
```

test/integration_test/my_tester/html/default_server/ファイル名.html

```
<h1>日本語でおk</h1>
```

TrainCoffeeBoy replied:
げえ、mergeした時に消えちゃったのか（飲酒運転ダメだ...)
足しておきます。

TrainCoffeeBoy replied:
笑, done!

dai65527 commented on test/unit_test/ParseRequest/test_parseRequest.cc:
```diff
@@ -760,38 +758,88 @@ TEST_F(test_parseRequest, RequestTesterPost2) {
+  EXPECT_EQ(request.headers_["transfer-encoding"], "chunked");
+  std::string str(request.body_.begin(), request.body_.end());
+  EXPECT_EQ("TRACE / HTTP/1.1\r\nHost:localhost\r\ntransfer-encoding:chunked\r\n\r\n3\r\n012\r\n0\r\n", str);
+}
```
改行

dai65527 commented on Request.cpp:
```diff
@@ -389,17 +397,21 @@ ssize_t Request::parseChunkedBody(size_t pos) {
+            std::vector<char>().swap(
+                buf_);  // free memory of buf_ (c11 can use fit_to_shurink);
+          }
+          return REQ_FIN_RECV;
```
この変更（おそらく）で42testerの後続のテストケースでエラーが出てしまうようです。バグの原因を探していますが、まだピンと来ません。

dai65527 reviewed (changes requested):
ご対応ありがとうございました！概ねLGTMです！
ただ、42Testerで0chunkedの後続のHEADがおかしくなっているように見えます。原因を探ってみていますが、まだピンと来ていません。

あと、テストケースの追加、ファイル末尾の改行をお願いします。

Commit:
decode case for integration test, add new line for unit_test

Commit:
html file for decode integration test

Commit:
merge remote origin

dai65527 committed:
mod: Request::parseChunkedBody

- `0\r\n\r` で終わっているなどの場合に内部で無限ループしていたのを解消
- chunk_size_ == 0の判定は\r\nの判定以降とした
- テストOK（単体テストは合わせて修正しています）

TrainCoffeeBoy committed:
Merge pull request #49 from dai65527/mod_request_decode_trace_dnakano

mod: Request::parseChunkedBody

Commit:
Merge branch 'mod_request_decode_trace' of github.com:dai65527/webserv into mod_request_decode_trace

Commit:
merge remote origin main

Commit:
testcase add

dai65527 reviewed (approved):
LGTMです！ありがとうございました！！！
